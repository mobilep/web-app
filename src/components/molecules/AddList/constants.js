export const keyMap = {
	ENTER: 13,
	UP_ARROW: 38,
	DOWN_ARROW: 40,
	TAB: 9,
};

export default {
	keyMap,
};
