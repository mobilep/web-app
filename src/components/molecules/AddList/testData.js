export const users = [
	{
		id: 1,
		fullName: 'John Smith',
		initials: 'JS',
	}, {
		id: 2,
		fullName: 'Philippe Starck',
		initials: 'PS',
	}, {
		id: 3,
		fullName: 'Suzy Salmon',
		initials: 'SS',
	},
];

export const allUsers = [
	{
		id: 4,
		fullName: 'John Rebele',
		initials: 'JR',
	}, {
		id: 5,
		fullName: 'John Klokdan',
		initials: 'JK',
	}, {
		id: 6,
		fullName: 'John Makeba',
		initials: 'JM',
	},
];
