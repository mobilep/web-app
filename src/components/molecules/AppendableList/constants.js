export const LIST_TYPES = {
	CHECK: 'check',
	STEP: 'step',
};

export default {
	LIST_TYPES,
};
