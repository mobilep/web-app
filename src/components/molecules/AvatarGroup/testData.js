export const avatars = [
	{
		firstName: 'Clayton',
		lastName: 'Gross',
		img: 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQBlUvtjoDyRYSmS6hoSVke1HcE-3ZeLGHLR7RnTIBp3jAYv39B',
		avatarColor: '#87d4ee',
	},
	{
		firstName: 'Clayton',
		lastName: 'Gross',
		img: 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQBlUvtjoDyRYSmS6hoSVke1HcE-3ZeLGHLR7RnTIBp3jAYv39B',
		avatarColor: '#87d4ee',
	},
	{
		firstName: 'Clayton',
		lastName: 'Gross',
		img: 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQBlUvtjoDyRYSmS6hoSVke1HcE-3ZeLGHLR7RnTIBp3jAYv39B',
		avatarColor: '#87d4ee',
	},
	{
		firstName: 'Clayton',
		lastName: 'Gross',
		img: 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQBlUvtjoDyRYSmS6hoSVke1HcE-3ZeLGHLR7RnTIBp3jAYv39B',
		avatarColor: '#87d4ee',
	},
	{
		firstName: 'Clayton',
		lastName: 'Gross',
		img: 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQBlUvtjoDyRYSmS6hoSVke1HcE-3ZeLGHLR7RnTIBp3jAYv39B',
		avatarColor: '#87d4ee',
	},
	{
		firstName: 'Clayton',
		lastName: 'Gross',
		img: 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQBlUvtjoDyRYSmS6hoSVke1HcE-3ZeLGHLR7RnTIBp3jAYv39B',
		avatarColor: '#87d4ee',
	},
];

export const avatarsLess = [
	{
		firstName: 'Clayton',
		lastName: 'Gross',
		img: 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQBlUvtjoDyRYSmS6hoSVke1HcE-3ZeLGHLR7RnTIBp3jAYv39B',
		avatarColor: '#87d4ee',
	},
	{
		firstName: 'Clayton',
		lastName: 'Gross',
		img: 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQBlUvtjoDyRYSmS6hoSVke1HcE-3ZeLGHLR7RnTIBp3jAYv39B',
		avatarColor: '#87d4ee',
	},
];
