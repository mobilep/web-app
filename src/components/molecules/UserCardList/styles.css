.UserCardList {
	padding: 0;
}

.UserCardList-item {
	list-style: none;
}
