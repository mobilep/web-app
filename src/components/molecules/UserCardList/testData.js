export const list = [
	{
		_id: '5aeb0146c9c09a7964d7631f',
		_recipient: {
			_id: '5ad862041addf72c9b7e69ff',
			firstName: 'Anton',
			lastName: 'Prok',
			avatarColor: 'fd524f',
			avatar_sm: null,
			havePassword: false,
			name: 'Anton Prok',
		},
		_user: {
			_id: '5aeafe9474f12e687e2c8a76',
			firstName: 'Kate',
			lastName: 'Vuilova',
			avatarColor: '6c6ced',
			avatar_sm: null,
			name: 'Kate Vuilova',
		},
		updatedAt: '2018-05-03T12:32:57.355Z',
		createdAt: '2018-05-03T12:32:06.298Z',
		inboxId: 'inboxes/5aeafe9474f12e687e2c8a76/5ad862041addf72c9b7e69ff/5aeb0146c9c09a7964d7631f',
		type: 'inbox',
		status: 'unread',
		message: {
			time: '1525350777342',
			_user: '5aeafe9474f12e687e2c8a76',
			content: "Hello, How you're getting on?",
			type: 'text',
			firebaseKey: '-LB_zR6F-9ur1dEyF6-t',
		},
		firstResponseAt: null,
		firstMessageAt: null,
		isActive: true,
	},
	{
		_id: '5aeafe94c9c09a7964d74809',
		_coach: {
			_id: '57fe2450916165b0b8b20111',
			firstName: 'Jon',
			lastName: 'Doe',
			avatarColor: '58c9ef',
			avatar_sm: null,
			name: 'Jon Doe',
		},
		_scenario: {
			_id: '574e3a50616165b0b8b55111',
			isActive: true,
			name: 'Demo Scenario',
			video: {
				playList: null,
				thumb: null,
			},
		},
		_user: {
			_id: '5aeafe9474f12e687e2c8a76',
			firstName: 'Kate',
			lastName: 'Vuilova',
			avatarColor: '6c6ced',
			lang: 'en',
			avatar_sm: null,
			name: 'Kate Vuilova',
		},
		updatedAt: '2018-05-03T12:20:36.194Z',
		__v: 0,
		createdAt: '2018-05-03T12:20:36.194Z',
		userEvaluatedAt: null,
		coachEvaluatedAt: null,
		isActive: true,
		status: 'current',
		coachMark: [],
		userMark: [],
		_inbox: {
			_id: '5aeafe9474f12e687e2c8a77',
			updatedAt: '2018-05-03T12:20:36.370Z',
			createdAt: '2018-05-03T12:20:36.178Z',
			_user: '5aeafe9474f12e687e2c8a76',
			_recipient: '57fe2450916165b0b8b20111',
			__v: 0,
			inboxId: 'chats/5a74f12e687e2c8a76/574e3a50616165b8b55111/5aeafe94c9c09a7964d74809/eafe9474f12e687e2c8a77',
			type: 'practice',
			status: 'unread',
			message: {
				time: '1525350036252',
				content: {
					thumb: 'https://s3-eu-west-1.amazonaws.com/mobilepractice-video/HLS/EN/thumbs/00001.png',
					playList: 'https://s3-eu-west-1.amazonaws.com/mobilepractice-video/HLS/EN/playlist.m3u8',
					videoOrientation: 'landscape',
					size: 394913852,
					duration: 34,
					videoId: 'EN',
				},
				_user: '57fe2450916165b0b8b20111',
				type: 'video',
				firebaseKey: '-LB_wbCKvKgs_eBmIepy',
			},
			firstResponseAt: null,
			firstMessageAt: null,
			isActive: true,
		},
	},
];

export const userList = [
	{
		_id: '5ac33a66b6c2a8602917029e',
		updatedAt: '2018-04-03T21:59:18.730Z',
		createdAt: '2018-04-03T08:25:10.366Z',
		firstName: 'Apple',
		lastName: 'Apple',
		_company: '5ac33a1eb6c2a8602917029d',
		avatarId: '8B613058-7CF1-4129-90AD-4A60AEC21671',
		firstLogIn: '2018-04-03T08:27:30.947Z',
		videosCount: 0,
		messagesCount: 10,
		isActive: true,
		avatarColor: 'f946e9',
		email: 'test+1006@techmagic.co',
		avatar_sm: 'https://s3-eu-west-1.amazonaws.com/mobilepractice-photo/public/100/8B613058-7CF1-4129-90AD-4A60AEC21671.jpg',
		avatar_md: 'https://s3-eu-west-1.amazonaws.com/mobilepractice-photo/public/640/8B613058-7CF1-4129-90AD-4A60AEC21671.jpg',
		avatar_lg: 'https://s3-eu-west-1.amazonaws.com/mobilepractice-photo/public/1024/8B613058-7CF1-4129-90AD-4A60AEC21671.jpg',
		name: 'Apple Apple',
	},
	{
		_id: '5c33a66b665860251702439',
		updatedAt: '2018-04-03T21:59:18.730Z',
		createdAt: '2018-04-03T08:25:10.366Z',
		firstName: 'Grape',
		lastName: 'Grape',
		_company: '5ac33a1eb6c2a8602917029d',
		avatarId: '8B613058-7CF1-4129-90AD-4A60AEC21671',
		firstLogIn: '2018-04-03T08:27:30.947Z',
		videosCount: 0,
		messagesCount: 10,
		isActive: true,
		avatarColor: 'f946e9',
		email: 'test+1006@techmagic.co',
		avatar_sm: 'https://s3-eu-west-1.amazonaws.com/mobilepractice-photo/public/100/81E557EC-D0CA-491C-971D-FC7B5CAB1B3C.jpg',
		name: 'Grape Grape',
	},
];
