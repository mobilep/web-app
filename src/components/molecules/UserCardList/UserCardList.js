import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import ConversationCard from '../ConversationCard';
import NewMessageCard from '../NewMessageCard';
import { CARD_TYPES } from '../UserCard/constants';
import { inboxListHelper } from '../../../helpers';
import constants from '../../../constants';
import './styles.css';
import { LanguageContext } from '../../../utils';

const { UNREAD } = constants.common.inboxStatuses;

export default class UserCardList extends Component {
	static contextType = LanguageContext;

	getPersonInitials (firstName = '', lastName = '') {
		return firstName.charAt(0) + lastName.charAt(0);
	}

	getListData ({
		listData,
		type,
		...rest
	}) {
		if (!listData) return [];

		if (type === CARD_TYPES.CONVERSATION_CARD) {
			return inboxListHelper({
				listData,
				...rest,
				...this.context.inbox,
			});
		}

		return listData;
	}

	handleCardSelect = (id, inboxList) => {
		this.props.onSelectCard(id, inboxList);
	}

	renderConversationList (listData) {
		const list = [...listData].sort((prevItem, item) => {
			if (item.message && prevItem.message) {
				const itemTime = Math.max(+item.message.time || 0, +new Date(item.createdAt));
				const prevItemTime = Math.max(+prevItem.message.time || 0, +new Date(prevItem.createdAt));
				try {
					const diff = itemTime - prevItemTime;
					if (diff) return diff;
				} catch (e) {
					return 0;
				}
			}
			return 0;
		});
		const { activeCard } = this.props;
		return list.map((item) => {
			const {
				_id,
				_recipient,
				message,
				updatedAt,
				status,
				_scenario,
				_inbox,
				state,
				groupChatName,
				groupAvatar,
			} = item;
			const personInitials = this.getPersonInitials(_recipient.firstName, _recipient.lastName);
			const isLastIncoming = _recipient._id === message._user;
			const coachEvaluated = item.coachMark && item.coachMark.length !== 0;
			const coachMark = coachEvaluated && item.coachMark.map((item) => item.mark);
			const coachMarkSum = coachMark && coachMark.reduce((a, b) => a + b, 0);
			const mark = coachMarkSum && coachMarkSum / coachMark.length;
			const evaluationMark = parseFloat(mark).toPrecision(2);
			let inboxStatus;

			if (status === UNREAD) {
				inboxStatus = status;
			} else {
				inboxStatus = _scenario ? _inbox.status : status;
			}
			return (
				<li className='UserCardList-item' key={_id} onClick={() => this.handleCardSelect(_id, listData)}>
					<ConversationCard
						waitingForFeedback={state === 'waitingOnFeedback'}
						active={_id === activeCard}
						avatar={groupAvatar || _recipient.avatar_sm}
						avatarColor={_recipient.avatarColor}
						date={+message.time || updatedAt}
						lastIncoming={isLastIncoming}
						message={message.content}
						person={_recipient.name || groupChatName}
						personInitials={personInitials}
						unread={inboxStatus === 'unread'}
						practiceCard={!!_scenario}
						evaluated={!!coachEvaluated}
						evaluationMark={evaluationMark} />
				</li>
			);
		});
	}

	getClassName = (className) => classNames('UserCardList', className);

	renderUserList (listData) {
		const { activeCard, inboxList } = this.props;

		return listData.map((item, index) => {
			const {
				_id,
				avatar_sm,
				avatarColor,
				firstName,
				lastName,
				name,
			} = item;
			const personInitials = this.getPersonInitials(firstName, lastName);

			return (
				<li className='UserCardList-item' key={_id} onClick={() => this.handleCardSelect(_id, inboxList)}>
					<NewMessageCard
						active={_id === activeCard}
						avatar={avatar_sm}
						avatarColor={avatarColor}
						person={name}
						personInitials={personInitials}
						first={index === 0}
					/>
				</li>
			);
		});
	}

	render () {
		const { type, className } = this.props;
		const listData = this.getListData(this.props);

		if (!listData || !listData.length) return null;

		return (<ul className={this.getClassName(className)}>
			{type === CARD_TYPES.CONVERSATION_CARD
				? this.renderConversationList(listData)
				: this.renderUserList(listData)
			}
		</ul>);
	}
}

UserCardList.propTypes = {
	activeCard: PropTypes.string,
	className: PropTypes.string,
	inboxList: PropTypes.array,
	listData: PropTypes.array,
	onComponentDidMount: PropTypes.func,
	onSelectCard: PropTypes.func,
	type: PropTypes.oneOf(Object.values(CARD_TYPES)),
};

UserCardList.defaultProps = {
	onComponentDidMount: () => {},
	onSelectCard: () => {},
};
