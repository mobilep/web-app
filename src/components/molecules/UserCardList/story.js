import React from 'react';
import { storiesOf } from '@storybook/react';

import UserCardList from './UserCardList';
import { CARD_TYPES } from '../../molecules/UserCard/constants';
import { userList, list } from './testData';

storiesOf('2. Molecules / UserCardList', module)
	.add('new message user list', () => (
		<UserCardList listData={userList} type={CARD_TYPES.NEW_MESSAGE_CARD} />
	))
	.add('conversation card list', () => (
		<UserCardList
			listData={list}
			type={CARD_TYPES.CONVERSATION_CARD} />
	));
