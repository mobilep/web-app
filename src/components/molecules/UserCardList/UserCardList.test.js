import React from 'react';
import { shallow } from 'enzyme';

import UserCardList from './UserCardList';
import { CARD_TYPES } from '../../molecules/UserCard/constants';
import { userList } from './testData';

describe('Molecule - UserCardList', () => {

	it('Should render', () => {
		const wrapper = shallow(
			<UserCardList
				listData={userList}
				type={CARD_TYPES.NEW_MESSAGE_CARD}
			/>
		);
		expect(wrapper).toMatchSnapshot();
	});

});
