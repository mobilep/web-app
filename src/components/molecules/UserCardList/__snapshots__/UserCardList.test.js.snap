// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molecule - UserCardList Should render 1`] = `
<ul
  className="UserCardList"
>
  <li
    className="UserCardList-item"
    key="5ac33a66b6c2a8602917029e"
    onClick={[Function]}
  >
    <NewMessageCard
      active={false}
      avatar="https://s3-eu-west-1.amazonaws.com/mobilepractice-photo/public/100/8B613058-7CF1-4129-90AD-4A60AEC21671.jpg"
      avatarColor="f946e9"
      first={true}
      person="Apple Apple"
      personInitials="AA"
    />
  </li>
  <li
    className="UserCardList-item"
    key="5c33a66b665860251702439"
    onClick={[Function]}
  >
    <NewMessageCard
      active={false}
      avatar="https://s3-eu-west-1.amazonaws.com/mobilepractice-photo/public/100/81E557EC-D0CA-491C-971D-FC7B5CAB1B3C.jpg"
      avatarColor="f946e9"
      first={false}
      person="Grape Grape"
      personInitials="GG"
    />
  </li>
</ul>
`;
